src/Database/LSMTree/Internal/Index.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE TypeFamilies #-}
+
+{-|
+    Provides a common interface to different types of fence pointer indexes and
+    their accumulators.
+-}
+module Database.LSMTree.Internal.Index
+(
+    Index (search, sizeInPages, fromSBS),
+    IndexAcc (ResultingIndex, appendSingle, appendMulti, unsafeEnd)
+)
+where
+
+import           Control.Monad.ST.Strict (ST)
+import           Data.ByteString.Short (ShortByteString)
+import           Data.Word (Word32)
+import           Database.LSMTree.Internal.Chunk (Chunk)
+import           Database.LSMTree.Internal.Entry (NumEntries)
+import           Database.LSMTree.Internal.Page (NumPages, PageSpan)
+import           Database.LSMTree.Internal.Serialise (SerialisedKey)
+
+-- | The class of index types.
+class Index i where
+
+    {-|
+        Searches for a page span that contains a key–value pair with the given
+        key. If there is indeed such a pair, the result is the corresponding
+        page span; if there is no such pair, the result is an arbitrary but
+        valid page span.
+    -}
+    search :: SerialisedKey -> i -> PageSpan
+
+    -- | Yields the number of pages covered by an index.
+    sizeInPages :: i -> NumPages
+
+    {-|
+        Reads an index along with the number of entries of the respective run
+        from a byte string.
+
+        The byte string must contain the serialised index exactly, with no
+        leading or trailing space.
+    -}
+    fromSBS :: ShortByteString -> Either String (NumEntries, i)
+
+{-|
+    The class of index accumulator types, where an index accumulator denotes an
+    index under incremental construction.
+
+    Incremental index construction is only guaranteed to work correctly when the
+    following conditions are met:
```

Comment:
Right. I didn’t move the corresponding documentation to the type class, because this requirement is specific to compact indexes. However, this restriction for compact indexes cannot be ignored when specifying the general interface, because this interface is also implemented for compact indexes. In the case of the requirement that the lengths of key sizes must fit into 16 bits each, I also moved the corresponding documentation to the type class. So, yes, I should do the same also with the requirement you mention here.

Reply:
Hmm, it actually seems that this requirement hadn’t even been documented before. I’ll add it then.

Reply:
The corresponding change is in eeec9da67dd9bbbdad34cb8d01213893becf3ffe.

Reply:
It was documented in `format-run.md`. https://github.com/IntersectMBO/lsm-tree/blob/11738bdd7b352d32e43cd8e45d93fde391c3b91d/doc/format-run.md?plain=1#L201

It's not explicitly mentioned in the code though, so it's good that we're mentioning it in the class documentation now